### ISSUE TITLE
Drawing bugs on Windows at very large scales

### ISSUE THREAD

**Issue by dannye** [fltk/fltk#1144]
**Describe the bug**

A little bit of background information on what I was trying to accomplish when I discovered these issues..

My application has a "background grid" with vertical lines. Some vertical lines are intentionally darker than others, but they are all 1 pixel thick.

When the window is scaled up, some of these lines are scaled up to 2 pixels thick. This is expected (not a bug at all), but very visually confusing and distracting in my use case:

https://github.com/user-attachments/assets/cb72bc28-ade6-4a62-aea5-836362d243ce

I found a pretty good solution for this (shown in the demo.cxx below) by using `fl_override_scale()` to ensure the vertical lines are always drawn with a width of 1, but it reveals some drawing bugs at scales >=200% which I have never seen under any other circumstances.

The bugs are:

1. These vertical lines still sometimes somehow get drawn with a thickness of 2.
2. The lines sometimes somehow get drawn on top of the children widgets that get drawn _after_ the lines.

**To Reproduce**

```cpp
#include <FL/Fl.H>
#include <FL/fl_draw.H>
#include <FL/Fl_Double_Window.H>
#include <FL/Fl_Scroll.H>
#include <FL/Fl_Group.H>
#include <FL/Fl_Box.H>

Fl_Double_Window *window = nullptr;

class Canvas : public Fl_Group {
public:
    Canvas(int x, int y, int w, int h, char *l=nullptr) : Fl_Group(x, y, w, h, l) {}

    void draw() {
        if (damage() & ~FL_DAMAGE_CHILD) {
            for (int i = 0; i < h() / 20; ++i) {
                fl_rectf(x(), y() + i * 20, w(), 20, i % 2 ? fl_gray_ramp(14) : fl_gray_ramp(10));
            }
            float scale = fl_override_scale();
            fl_push_clip(
                parent()->x() * scale + 0.001f,
                parent()->y() * scale + 0.001f,
                (parent()->w() - Fl::scrollbar_size()) * scale + 0.001f,
                (parent()->h() - Fl::scrollbar_size()) * scale + 0.001f
            );
            for (int i = 0; i < w() / 50; ++i) {
                fl_color(i % 4 ? fl_gray_ramp(2) : FL_BLACK);
                int x_pos = x() + i * 50;
                fl_yxline(
                    (int)((x_pos + 1)       * scale + (x_pos + 1       < 0 ? -0.001f : 0.001f)) - 1,
                    (int)(y()               * scale + (y()             < 0 ? -0.001f : 0.001f)),
                    (int)(((y() + h()) + 1) * scale + ((y() + h() + 1) < 0 ? -0.001f : 0.001f)) - 1
                );
            }
            fl_pop_clip();
            fl_restore_scale(scale);
        }
        draw_children();
    }

    int handle(int event) {
        if (event == FL_PUSH) {
            redraw();
            return 1;
        }
        return Fl_Group::handle(event);
    }
};

int main(int argc, char **argv) {
    window = new Fl_Double_Window(200, 200, 400, 400);
    Fl_Scroll *scroll = new Fl_Scroll(50, 50, 300, 300);
    Canvas *canvas = new Canvas(50, 50, 3000, 3000);
    Fl_Box *box = new Fl_Box(FL_BORDER_BOX, 100, 210, 200, 20, "");
    box->color(FL_RED);
    window->end();

    window->size_range(200, 200);
    window->resizable(scroll);

    window->show(argc, argv);
    return Fl::run();
}
```

1. Build and run the above demo.cxx like normal.
3. Scale the window up to 200% or greater.
4. Click on the grid with the mouse (to trigger a redraw).
5. Bug 1: See that some of the lines now have a thickness of 2. (You may need to resize the window, or scale it up or down, or scroll around, and repeat step 3; because this effect is not very predictable.)
6. Scroll the scrollbars left/right or up/down
7. Bug 2: See that the vertical lines are now drawn on top of the red box towards the top-left of the scroll area.
8. Repeat the steps at 170% scale and see that neither of the bugs occur.

**Expected behavior**

If `fl_override_scale()` is used followed by `fl_yxline()`, then the line should have a width of 1 regardless of the screen scaling.
If `fl_override_scale()` is used followed by `fl_yxline()` to draw lines in a Fl_Group subclass's `draw()` function before the children are drawn, then the lines should not appear on top of the children regardless of the screen scaling,

**Video**

https://github.com/user-attachments/assets/baa294e9-cf3f-4572-a9bb-9c43a6967f21

**FLTK Version**
 - Version: 1.4.0
 - Downloaded and self-built from: Git
 - If from Git, branch: master
 - If from Git, commit: 801125b

**FLTK Configure / Build Options**
 - Default VS CMake options

**Operating System / Platform:**
 - OS: Windows
 - OS Version: Windows 11
 - Processor if relevant: Intel

**Comment by dannye** [comment]
X11 without Cairo also has Bug 2 (lines on top of children at scales >=200%).

X11 with Cairo has Bug 1 and Bug 2, and also has another issue where the clipping region doesn't work as expected when `fl_override_scale()` is called and doesn't restore correctly when `fl_pop_clip()` and `fl_restore_scale(scale)` are called:

![image](https://github.com/user-attachments/assets/ca43ba75-b7f9-435f-ade5-a71a30f10a96)

I can submit a separate issue for this if needed.

**Comment by ManoloFLTK** [comment]
@dannye I believe you are trying to ask too much from the FLTK drawing API which uses integer-valued coordinates and therefore cannot do things expected from a real-valued drawing system. FLTK 1.4 allows to scale up (not to scale below 1) what is drawn at scale 1 with the 1.3 API keeping graphics generally proportional to their result at scale 1, but with some non proportional details due to the API being integer-valued. Line width is one of these details. It goes from 1 to 2 when the scaling factor is 2 or more. Drawing lines of width 1 with a scaling factor or 2 is beyond what is possible with the public API.

I believe another approach is to be used to do what you want, draw thin lines even with an apparent high scale factor. You should draw at FLTK scale 1 to an offscreen surface (that is, an `Fl_Image_Surface` with the default value - 0 - of the `Fl_Image_Surface` constructor's 3rd argument) and put the scaling logic in your user code where you can draw longer lines and keep them thin. Finally use `fl_copy_offscreen()` to copy the offscreen graphics to your canvas on display. See the `test/offscreen` demo program for a complete example of this approach. Member function `Fl_Image_Surface::offscreen()` returns the `Fl_Offscreen` object needed for the 5th argument of `fl_copy_offscreen()`.

**Comment by dannye** [comment]
> but with some non proportional details due to the API being integer-valued. Line width is one of these details. It goes from 1 to 2 when the scaling factor is 2 or more.

Okay. This is a little hard for me to understand because I am only trying to use native pixel, 1-to-1 integer coordinates, not real coordinates, and I am trying to draw at a 1x scaling factor and not a 2x scaling factor. But I of course acknowledge that my understanding of the technical details of FLTK are far, far inferior to yours.

My goals are very much the same as the OP's goals from this discussion thread where I learned about `fl_override_scale()`: https://groups.google.com/g/fltkgeneral/c/gcqn9EeA7_A/m/f6hSX4c3AgAJ

Based on that discussion, my understanding of `fl_override_scale()` is that it _reverses the affects of scaling_ so it's surprising to learn that there are caveats to that.

For now, another compromise I will consider is: Since my goal primarily is to avoid the "uneven" grid lines, I will try to only apply the override to "enforce thin lines" when the screen scale is between 100% and 200%. And at 200%, just let all the vertical lines be 2 pixels wide. What's most important is that they're all the same.

For the X11+Cairo bug where `fl_override_scale()` and `fl_push_clip()` do not cooperate, I will try to simplify the example further and submit a separate issue.

I'd request that this issue be left open as a RFE for greater control over drawing, even if no action is taken for the foreseeable future. I think this RFE is important for creating truly DPI-aware widgets, as intended by the OP of the discussion thread.

**Comment by ManoloFLTK** [comment]
@dannye This issue should be fixed after the commit mentioned above.

**Comment by dannye** [comment]
@ManoloFLTK 

This is the result I get with the above demo program on Windows:

https://github.com/user-attachments/assets/d9ecfba8-7092-40cd-b328-b9977d1486af

And on X11 without Cairo:

https://github.com/user-attachments/assets/97d5f113-b176-4eaf-b711-e937cf1effbf

And on X11 with Cairo:

https://github.com/user-attachments/assets/e220a1af-7b50-42ae-ae0f-c81b4833927c

I'm currently not sure if this is because the demo program needs to change how it uses `fl_push_clip()`/`fl_pop_clip()`. But currently, the result is very different than the second video in the OP.
If you could advise on if the demo needs changes in order to override the scale and manage the clip correctly, I would appreciate it. Thanks.

**Comment by dannye** [comment]
Oh, sorry, I just now tried simply removing `fl_push_clip()` and `fl_pop_clip()` and that seems to be what is needed here.

I can confirm Bug 2 is fixed (lines drawn on top of children at >=200% scale), at least on Windows. I'll need to check X11 again.

Bug 1 seems unaffected (sometimes thicker lines at >=200% scale). But still, this is pretty minor.

**Comment by dannye** [comment]
> I'll need to check X11 again.

All issues on X11 (with and without Cairo) are resolved.

**Comment by ManoloFLTK** [comment]
Yes, I forgot to mention that the `fl_push_clip()` call needs to be moved out of the `fl_override_scale()/fl_restore_scale()` section. Thus, an `fl_push_clip()` can still be present but transformed into 
`fl_push_clip( parent()->x(), parent()->y(), parent()->w() - Fl::scrollbar_size(), parent()->h() - Fl::scrollbar_size() );` and located just before the `fl_override_scale()` call.

**Comment by ManoloFLTK** [comment]
@dannye  Bug 1 also should be fixed after e9ea177.

**Comment by dannye** [comment]
Thanks! Everything looks really great, I really appreciate it!
